### ISSUE TITLE
Array count operator's return doesn't alias original elements

### ISSUE THREAD

**Issue by cassella** [chapel-lang/chapel#13217]
### Summary of Problem

The spec section 27.1 describing the Array Count operator says

```
The # operator can be applied to dense rectangular arrays with a tuple
argument whose size matches the rank of the array (or optionally an
integer in the case of a 1D array). The operator is equivalent to
applying the # operator to the array’s domain and using the result to
slice the array as described in Section 21.6.1.
```

But I find it's not equivalent.  Iterating over such a slice yields
refs to the array's elements as normal, but iterating over `A#count`
does not, AFAICT.

```chapel
const D = {1..10};

var BA, BB, BC: [D] int;

writeln("iterate over domain-count");
for ba in BA.domain#5 do
  BA[ba] = 1;
writeln(BA);

writeln("iterate over slice by counted domain");
var DC = BB.domain#5;
for ba in BB[DC] do
  ba = 1;
writeln(BB);

writeln("iterate over array-count");
for ba in BC#5 do
  ba = 1;
writeln(BC);

writeln("writeln of array-count");
BC = 1..10;
writeln(BC#5);

writeln("iterate over AA = BC#5");
var AA = BC#5;
for ba in AA do
  ba = 1;
writeln("AA: ", AA);
writeln("BC: ", BC);
```

```
iterate over domain-count
1 1 1 1 1 0 0 0 0 0
iterate over slice by counted domain
1 1 1 1 1 0 0 0 0 0
iterate over array-count
0 0 0 0 0 0 0 0 0 0
writeln of array-count
1 2 3 4 5
iterate over AA = BC#5
AA: 1 1 1 1 1
BC: 1 2 3 4 5 6 7 8 9 10
```

The writeln of array-count shows that iterating over `A#count` at
least yields the values of the array elements.

The `AA = BC#5` example I think shows that `BC#5` is not returning an
array whose elements are aliases of `BC`'s, per 21.6.1 Rectangular
Array Slicing.

**Associated Future Test(s):**

None yet.

### Configuration Information

```
chpl version 1.20.0 pre-release (470008e)

CHPL_TARGET_PLATFORM: linux64
CHPL_TARGET_COMPILER: gnu
CHPL_TARGET_ARCH: x86_64
CHPL_TARGET_CPU: native *
CHPL_LOCALE_MODEL: flat
CHPL_COMM: gasnet *
  CHPL_COMM_SUBSTRATE: udp
  CHPL_GASNET_SEGMENT: everything
CHPL_TASKS: qthreads
CHPL_LAUNCHER: amudprun
CHPL_TIMERS: generic
CHPL_UNWIND: none
CHPL_MEM: jemalloc
CHPL_ATOMICS: cstdlib
  CHPL_NETWORK_ATOMICS: none
CHPL_GMP: gmp
CHPL_HWLOC: hwloc
CHPL_REGEXP: re2
CHPL_LLVM: none
CHPL_AUX_FILESYS: none

gcc (Ubuntu 5.4.0-6ubuntu1~16.04.11) 5.4.0 20160609
```

**Comment by cassella** [comment]
My motivation was to explore replacing this loop in
`modules/packages/Sort.chpl` `isSorted()`

```chapel
forall (element, i) in zip(Data, Dom) with (&& reduce sorted) {
    if i > Dom.low {
      sorted &&= (chpl_compare(Data[i-stride], element, comparator) <= 0);
    }
}
```

with something like this loop, which doesn't have to worry about the
stride or alignment, or have a special case for one iteration:

```chapel
forall (a, b) in zip(Data#(Dom.size-1), Data#-(Dom.size-1)) with (&& reduce sorted) {
    sorted &&= (chpl_compare(a, b, comparator) <= 0);
}
```

But I ran into this issue trying to verify that this loop does the
right thing, locality-wise:

`forall a in A#n do a = here.id`

Using the `var AA = BC#n` approach, I believe that it does.

**Comment by bradcray** [comment]
Thanks for asking about this.  I can never even remember whether or not `#` was implemented for arrays.

Looking at the implementation of `#` on arrays, it appears to be doing the correct thing, but I think does have a subtle bug and is returning a deep-copy of the array rather than a shallow copy.  Looks like it should be a super-quick fix.

That said, I think there's also a bug in your code and that you'd need to write `ref AA = BC#5;` to get the behavior you were looking for because otherwise `AA` will be doing a deep copy of `BC#5` as it creates a new variable of its type (?).  This is consistent with other forms of slicing arrays.

**Comment by cassella** [comment]
Oops, thanks!

**Comment by bradcray** [comment]
This was as easy to fix as hoped:  See PR #13223.
